### PR TITLE
Allow unaligned load/store in QVM interpreter/x86 compiler

### PR DESCRIPTION
**Pull request by ec-**
Constructions like (dataMask & ~3) was used to protect against out-of-bound load/store when address is 4-byte closer to dataMask.
But at the same time it effectively cut low address bits for ALL load/store operations which is totally wrong in terms of conformance to ALLOWED (i.e. generated by q3lcc from C sources) low-level operations like packed binary data parsing

### CODE REVIEW

**Comment by ec-** [comment]
Here is simple test code that behaves differently in current interpreter/compiler:
`void CG_Test_f( void ) {
	static char data[8] = { 0x0, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7 };
	short *s;
	int *d;
	s = (short*)&data[3];
	d = (int*)&data[3];
	if ( (*s == 0x0304 || *s == 0x0403) && (*d == 0x03040506 || *d == 0x06050403 ) )
		CG_Printf( "your QVM is fine: short=%i, int=%i", *s, *d );
	else
		CG_Printf( "your QVM is broken: short=%i, int=%i", *s, *d );
}`

File: code/qcommon/vm.c [hunk 1]
```diff
@@ -451,13 +451,15 @@ vmHeader_t *VM_LoadQVM( vm_t *vm, qboolean alloc, qboolean unpure)
 	if(alloc)
 	{
 		// allocate zero filled space for initialized and uninitialized data
-		vm->dataBase = Hunk_Alloc(dataLength, h_high);
+		// leave some space beyound data mask so we can secure all mask operations
```
Comment (timangus) [inline code/qcommon/vm.c]: beyound?

**Comment by timangus** [comment]
The test code for people that aren't autistic savants:
`void CG_Test_f(void)
{
  static char data[8] = { 0x0, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7 };
  short *s;
  int *d;
  s = (short*)&data[3];
  d = (int*)&data[3];
  if((*s == 0x0304 || *s == 0x0403) && (*d == 0x03040506 || *d == 0x06050403))
    CG_Printf("your QVM is fine: short=%i, int=%i", *s, *d);
  else
    CG_Printf("your QVM is broken: short=%i, int=%i", *s, *d);
}
`

**Comment by timangus** [comment]
```void CG_Test_f(void)
{
  static char data[8] = { 0x0, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7 };
  short *s;
  int *d;
  s = (short*)&data[3];
  d = (int*)&data[3];
  if((*s == 0x0304 || *s == 0x0403) && (*d == 0x03040506 || *d == 0x06050403))
    CG_Printf("your QVM is fine: short=%i, int=%i", *s, *d);
  else
    CG_Printf("your QVM is broken: short=%i, int=%i", *s, *d);
}```

**Comment by ensiform** [comment]
\`\`\`cpp
\<code\>
\`\`\`

Will give syntax highlighting too. And afaik as a maintainer @timangus should be able to edit the initial comment.

**Comment by timangus** [comment]
Are there any negative effects to applying this?

**Comment by ec-** [comment]
@timangus No, because additional 4 bytes is allocated at end of data segment to handle possible out-of-bounds access (line 455 and 456). There also slight performance increase for interpreter

**Comment by timangus** [comment]
Ahh right, I had completely glossed over that.

**Pull request merged**